File: topk-rs/tests/test_query_filter.rs
use test_context::test_context;
use topk_protos::v1::data::{stage::filter_stage::FilterExpr, LogicalExpr, Query, Stage};

mod utils;
use utils::dataset;
use utils::ProjectTestContext;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_starts_with(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            Query::new(vec![
                Stage::filter(FilterExpr::logical(LogicalExpr::starts_with(
                    LogicalExpr::field("_id"),
                    LogicalExpr::literal("cat".into()),
                ))),
                Stage::topk(LogicalExpr::field("published_year"), 100, false),
            ]),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert_doc_ids!(result, ["catcher"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_starts_with_empty(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            Query::new(vec![
                Stage::filter(FilterExpr::logical(LogicalExpr::starts_with(
                    LogicalExpr::field("_id"),
                    LogicalExpr::literal("".into()),
                ))),
                Stage::topk(LogicalExpr::field("published_year"), 100, false),
            ]),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert_doc_ids!(
        result,
        [
            "gatsby",
            "catcher",
            "moby",
            "mockingbird",
            "alchemist",
            "harry",
            "lotr",
            "pride",
            "1984",
            "hobbit"
        ]
    );
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_starts_with_non_existent_prefix(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            Query::new(vec![
                Stage::filter(FilterExpr::logical(LogicalExpr::starts_with(
                    LogicalExpr::field("_id"),
                    LogicalExpr::literal("foobarbaz".into()),
                ))),
                Stage::topk(LogicalExpr::field("published_year"), 100, false),
            ]),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(result.is_empty());
}

File: topk-py/tests/test_query_filter.py
from topk_sdk.query import field, filter

from . import ProjectContext
from .utils import dataset


def test_query_filter_basic(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    results = ctx.client.collection(collection.name).query(
        filter(field("published_year") <= 1950).topk(
            field("published_year"), 100, False
        )
    )

    assert [doc["_id"] for doc in results] == [
        "1984",
        "hobbit",
        "gatsby",
        "moby",
        "pride",
    ]


def test_query_filter_with_lte_operator(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    results = ctx.client.collection(collection.name).query(
        filter(field("published_year").lte(1950)).topk(
            field("published_year"), 100, False
        )
    )

    assert [doc["_id"] for doc in results] == [
        "1984",
        "hobbit",
        "gatsby",
        "moby",
        "pride",
    ]


def test_query_filter_with_gte_operator(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    results = ctx.client.collection(collection.name).query(
        filter(field("published_year") >= 1950).topk(field("published_year"), 100, True)
    )

    # The actual count depends on the dataset, so we'll check the count and some key documents
    assert len(results) == 5
    doc_ids = [doc["_id"] for doc in results]
    assert "catcher" in doc_ids
    assert "1984" in doc_ids


def test_query_filter_with_eq_operator(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    results = ctx.client.collection(collection.name).query(
        filter(field("published_year") == 1951).topk(field("published_year"), 100, True)
    )

    assert [doc["_id"] for doc in results] == ["catcher"]


def test_query_filter_with_ne_operator(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    results = ctx.client.collection(collection.name).query(
        filter(field("published_year") != 1951).topk(field("published_year"), 100, True)
    )

    assert len(results) == 9
    assert "catcher" not in [doc["_id"] for doc in results]
